pkg/services/store/entity/sqlstash/broadcaster.go

```diff
@@ -23,101 +102,132 @@ func NewBroadcaster[T any](ctx context.Context, connect ConnectFunc[T]) (Broadca
+	select {
+	case <-ctx.Done(): // client canceled
+		return nil, ctx.Err()
+	case <-b.terminated: // no more data
+		return nil, io.EOF
+	case b.subscribe <- sub: // success submitting subscription
+		return sub, nil
+	}
 }
 
-func (b *broadcaster[T]) Unsubscribe(sub chan T) {
-	b.unsubscribe <- sub
-}
+func (b *broadcaster[T]) Unsubscribe(sub <-chan T) {
+	// wait for broadcaster to start. In practice, the only way to reach
+	// Unsubscribe is by first having called Subscribe, which means we have
+	// already started. But a malfunctioning caller may call Unsubscribe freely,
+	// which would cause us to block forever the goroutine of the caller when
+	// trying to send to a nil `b.unsubscribe` or receive from a nil
+	// `b.terminated` if we haven't yet initialized those values. This would
+	// mean leaking that malfunctioninig caller's goroutine, so we rather make
+	// Unsubscribe safe in any possible case
+	if sub == nil {
+		return
+	}
+	<-b.started // wait for broadcaster to start
 
-func (b *broadcaster[T]) start(ctx context.Context, connect ConnectFunc[T]) error {
-	if b.running {
-		return fmt.Errorf("broadcaster already running")
+	select {
+	case b.unsubscribe <- sub: // success submitting unsubscription
+	case <-b.terminated: // broadcaster terminated, nothing to do
 	}
+}
 
+// init initializes the broadcaster. It should not be run more than once.
+func (b *broadcaster[T]) init(ctx context.Context, connect ConnectFunc[T]) error {
+	// create the stream that will connect us with the watch implementation and
+	// send it to them so they initialize and start sending data
 	stream := make(chan T, 100)
-
-	err := connect(stream)
-	if err != nil {
+	if err := connect(stream); err != nil {
 		return err
 	}
 
-	b.ctx = ctx
-
+	// initialize our internal state
+	b.shouldTerminate = ctx.Done()
 	b.cache = NewCache[T](ctx, 100)
 	b.subscribe = make(chan chan T, 100)
-	b.unsubscribe = make(chan chan T, 100)
-	b.subs = make(map[chan T]struct{})
+	b.unsubscribe = make(chan (<-chan T), 100)
+	b.subs = make(map[<-chan T]chan T)
+	b.terminated = make(chan struct{})
 
+	// start handling incoming data from the watch implementation. If data came
+	// in until now, it will be buffered in `stream`
 	go b.stream(stream)
 
-	b.running = true
+	// unblock any Subscribe/Unsubscribe calls since we are ready to handle them
+	close(b.started)
+
 	return nil
 }
 
-func (b *broadcaster[T]) stream(input chan T) {
+// stream acts a message broker between the watch implementation that receives a
+// raw stream of events and the individual clients watching for those events.
+// Thus, we hold the receive side of the watch implementation, and we are
+// limited here to receive from it, whereas we are responsible for sending to
+// watchers and closing their channels. The responsibility of closing `input`
+// (as with any other channel) will always be of the sending side. Hence, the
+// watch implementation should do it.
+func (b *broadcaster[T]) stream(input <-chan T) {
```

Comment:
Changed the signature to receive a `<-chan T` instead of `chan T`. Within the method, we no longer close `input`. That is moved to the write path (in `poller`). Please, read the extensive comment about how channels work to understand why it's not good to mix the read and write path of a channel (TL;DR better channel hygiene, separate producer and consumer concerns, prevent panics if not carefully handled).